runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/ThreadJob.java

```diff
@@ -317,11 +323,16 @@ private static ThreadJob waitForRun(final ThreadJob threadJob, IProgressMonitor
 						// The actual exit conditions are listed above at the beginning of
 						// this while loop
 						int state = blockingJob.getState();
-						//ensure we don't wait forever if the blocker is waiting, because it might have yielded to me
-						if (state == Job.RUNNING && canBlock) {
-							blockingJob.jobStateLock.wait();
-						} else if (state != Job.NONE) {
-							blockingJob.jobStateLock.wait(250);
+						// Check that blockingJob has not acquired a different, non-conflicting
+						// scheduling rule since checking for conflicts by JobManager
+						if (state != Job.NONE && blockingJob.getRule().isConflicting(threadJob.getRule())) {
+							// ensure we don't wait forever if the blocker is waiting, because it might have
+							// yielded to me
+							if (state == Job.RUNNING && canBlock) {
+								blockingJob.jobStateLock.wait();
```

Comment:
there is still no guarantee that the rule may change between the isConflicting() check and this wait.

Reply:
I am not sure whether I understand your comment correcly.

You mean that the rule may _not_ have changed since the added `isConflicting()` call in line 328?
Correct me if I'm wrong, but since this code block is synchronized on `blockingJob.jobStateLock`, the rule of the `blockingJob` cannot change between the `isConflicting()` check and the `wait()` call. At least, that is exactly what the the `jobStateLock` is supposed to be used for.
And the rule of `threadJob` cannot change either, as that is the job is currently not executed at all.

In case you refer to the existing conflict check performed in `runNow()`, then in case this `wait()` is reached, the state did actually _not_ change. That is exactly why this `wait()` can be called safely, as a `notify()` will wake up the thread on a state change of the other job.

Maybe you can clarify in which scenario the behavior is incorrect.

Reply:
The synchronize ensures that bolingJob.getState() can not change. but i don't see a guarantee that the result of "blockingJob.getRule().isConflicting(threadJob.getRule())" can not change during the synchronized block. The ISchedulingRule.isConflicting(ISchedulingRule) can be custom code doing anything.

Reply:
or numberOfParallelBuilds = 300 this PR seems to have the same effect as simply changing .wait() to .wait(250) . May be it would be better to either just never unconditionally wait, or to find the right spot to notify the synchronizer.

Reply:
> The synchronize ensures that bolingJob.getState() can not change. but i don't see a guarantee that the result of "blockingJob.getRule().isConflicting(threadJob.getRule())" can not change during the synchronized block. The ISchedulingRule.isConflicting(ISchedulingRule) can be custom code doing anything.

Okay, I got your point, thanks! However, `ISchedulingRule.isConflicting(ISchedulingRule)` cannoit to anything, as its specification says that "Implementations of this method must be reflexive, symmetric, and consistent". So for the same pair of rules, calling `isConflicting()` multiple times has to yield the same value.

> for numberOfParallelBuilds = 300 this PR seems to have the same effect as simply changing .wait() to .wait(250) . May be it would be better to either just never unconditionally wait, or to find the right spot to notify the synchronizer.

An unconditional wait would also solve the problem and maybe you are right that we should never wait unconditionally. Still I would be in favor to check the conflicting state again before waiting, as the wait is unnecessary in case the conflict got resolved meanwhile.